File: google_news.py
# -*- coding: utf-8 -*-

import urllib
import urllib2

class GoogleNews:
  @classmethod
  def create_params(cls, keyword):
    return {
      "hl"    : "ja",
      "ned"   : "us",
      "ie"    : "UTF-8",
      "oe"    : "UTF-8",
      "output": "rss",
      "q"     : keyword.encode("utf-8"),
    }

  @classmethod
  def create_url(cls, keyword):
    params = cls.create_params(keyword)
    return "http://news.google.com/news?" + urllib.urlencode(params)

  @classmethod
  def fetch_rss(cls, keyword):
    url = GoogleNews.create_url(keyword)
    request = urllib2.Request(url = url)
    request.add_header("User-Agent", "ironnews")
    io = urllib2.urlopen(request)
    try:
      return io.read()
    finally:
      io.close()

src = GoogleNews.fetch_rss(u"鉄道")
print src

File: ironnews-helper2/lib/google_news.rb

require "cgi"

module GoogleNews
  def self.create_params(options = {})
    options = options.dup
    keyword = options.delete(:keyword) || raise(ArgumentError)
    num     = options.delete(:num) || 10
    raise(ArgumentError) unless options.empty?

    return {
      "hl"     => "ja",
      "ned"    => "us",
      "ie"     => "UTF-8",
      "oe"     => "UTF-8",
      "output" => "rss",
      "num"    => num.to_s,
      "q"      => keyword,
    }
  end

  def self.create_base_url
    return "http://news.google.com/news"
  end

  def self.create_url(options = {})
    options = options.dup
    keyword = options.delete(:keyword) || nil
    num     = options.delete(:num)     || nil
    raise(ArgumentError) unless options.empty?

    query = self.create_params(:keyword => keyword, :num => num).
      sort_by { |key, value| key }.
      map     { |key, value| CGI.escape(key) + "=" + CGI.escape(value.to_s) }.join("&")
    return self.create_base_url + "?" + query
  end
end
